\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in,marginparwidth=2in]{geometry}

% use Unicode characters - try changing the option if you run into troubles with special characters (e.g. umlauts)
\usepackage[utf8]{inputenc}

% clean citations
\usepackage{cite}

% hyperref makes references clicky. use \url{www.example.com} or \href{www.example.com}{description} to add a clicky url
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% improves typesetting in LaTeX
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% text layout - change as needed
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in 
\textheight 8.75in

% Remove % for double line spacing
%\usepackage{setspace} 
%\doublespacing

% use adjustwidth environment to exceed text width (see examples in text)
\usepackage{changepage}

% adjust caption style
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,singlelinecheck=off]{caption}

% remove brackets from references
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% headrule, footrule and page numbers
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}

% use \textcolor{color}{text} for colored text (e.g. highlight to-do areas)
\usepackage{color}

% define custom colors (this one is for figure captions)
\definecolor{Gray}{gray}{.25}

% this is required to include graphics
\usepackage{graphicx}

% use if you want to put caption to the side of the figure - see example in text
\usepackage{sidecap}

% use for have text wrap around figures
\usepackage{wrapfig}
\usepackage[pscoord]{eso-pic}
\usepackage[fulladjust]{marginnote}
\reversemarginpar

% Adding multirow.
\usepackage{multirow}

% Other required things:
\usepackage{subcaption}
\captionsetup[subfigure]{justification=centering}
\usepackage{xcolor}
\usepackage[printwatermark]{xwatermark}
\newwatermark[allpages,color=gray!20,angle=45,scale=3,xpos=0,ypos=0]{CONFIDENTIAL}

% document begins here
\begin{document}
\vspace*{0.35in}

% title goes here:
\begin{flushleft}
{\Large
    \textbf\newline{Avoiding illusions in single cell RNA sequencing data analysis}
}
\newline

% authors go here:
%\\
Aaron Lun\textsuperscript{1,*}
\\
\bigskip
\bf{1} Cancer Research UK Cambridge Institute, University of Cambridge, Li Ka Shing Centre, Robinson Way, Cambridge CB2 0RE, United Kingdom \\
\bigskip

\end{flushleft}

\noindent
A new imputation method ``MAGIC" \cite{vandijk2018recovering} has recently been proposed for denoising single-cell RNA sequencing (scRNA-seq) data.
This method involves constructing an affinity matrix based on the distances between cells in high-dimensional space;
applying a diffusion process to strengthen similarities between cells due to existing trends in the data;
and multiplying the exponentiated affinity matrix with the original count matrix to obtain denoised expression values for downstream analysis.
The authors claim that the use of MAGIC removes technical noise and improves the biological signal in scRNA-seq data.
Indeed, the noisy nature of scRNA-seq data has been documented \cite{grun2015design} and many computational methods have been developed to address this problem \cite{bacher2016design}.
Unfortunately - as we will argue below - using MAGIC is not the solution.

The raison d'\^etre of the MAGIC procedure is the use of the diffusion process to reinforce existing trends.
The affinity matrix is exponentiated to represent the probability of one cell reaching another via a random walk of length $t$, possibly involving a number of intermediate cells.
This procedure systematically increases the affinities between pairs of cells that are connected by many close neighbours.
Conversely, affinities are downweighted between pairs of cells in sparse regions of the high dimensional space, 
where sporadic similarities are more likely to be caused by technical noise or outliers.
By performing diffusion, MAGIC can achieve aggressive denoising compared to other methods like $k$-NN imputation.
However, this also runs the risk of distorting the biological structure in the denoised expression matrix.

This effect is best demonstrated with a simple simulation.
We simulated a matrix of Poisson-distributed counts that contained no structure other than modest differences in library size across cells (Figure~\ref{fig:nostructure}a).
Upon applying MAGIC and performing a principal components analysis (PCA) on the denoised expression matrix, we observed a ``trajectory'' of cells along the first principal component (PC) (Figure~\ref{fig:nostructure}b) that was not present with the original expression matrix.
This trajectory is a manifestation of spurious structure, likely arising from stochastic differences in the affinity matrix that are inflated during diffusion.
These initial differences in affinity are probably driven by differences in variance when library sizes are variable across cells, assuming that mean differences are eliminated by library size normalization.
While the creation of artificial structure is obviously undesirable, the interpretation of denoised expression values at the gene level is even more compromised.
All genes are known to be independent in this simulation (Figure~\ref{fig:nostructure}c) but now appear to be strongly co-regulated (Figure~\ref{fig:nostructure}d).

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 15mm,clip,page=3]{../scripts/pics/nostruct.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 15mm,clip,page=1]{../scripts/pics/nostruct.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 15mm,clip,page=4]{../scripts/pics/nostruct.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 15mm,clip,page=2]{../scripts/pics/nostruct.pdf}
    \caption{}
\end{subfigure}
\caption{Effects of applying MAGIC to a simulated dataset containing independently sampled Poisson-distributed counts with no structure.
(a) Plot of the first two PCs after performing PCA on the log-normalized expression matrix.
Each point represents a cell, coloured according to its library size (purple-green-yellow gradient, from lowest to highest).
The percentage of variance explained by each PC is shown in parentheses.
(b) Same as (a) but on the denoised expression values after running MAGIC.
(c) Log-expression values of the two genes with the highest pairwise correlation.
Each point represents a cell.
The line of best fit and R$^2$ are shown in red.
(d) Same as (c), for the most correlated gene pair with denoised expression values after running MAGIC.}
\label{fig:nostructure}
\end{figure}

One might dismiss the previous simulation as being somewhat pathological.
After all, there are no biological systems that exhibit no structure whatsoever.
Perhaps a more pertinent question would be: how does MAGIC perform in the presence of genuine structure? 
To explore this, we performed another simulation involving four equally sized subpopulations of cells, denoted here as $A$, $B$, $C$ and $D$.
The primary separation occurs between $\{A, B\}$ and $\{C, D\}$ and the secondary separation occurs between $\{A, C\}$ and $\{B, D\}$ (Figure~\ref{fig:fourclusters}a).
Application of MAGIC eliminated the secondary separation (Figure~\ref{fig:fourclusters}b), merging $A$ with $B$ and $C$ with $D$.
This is again a consequence of the diffusion process, where weak similarities between cells in $A$ and $B$ (or $C$ and $D$) are increasingly strengthened over diffusion time.
More information is subsequently shared between those pairs of subpopulations during the denoising step, squeezing them together and eliminating geniune structure.
Gene-level interpretation is also compromised as differences in expression between subpopulations are shrunk towards zero for the subset of genes responsible for the secondary separation (Figure~\ref{fig:fourclusters}c, d).

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=3]{../scripts/pics/structloss.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{../scripts/pics/structloss.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=4]{../scripts/pics/structloss.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=2]{../scripts/pics/structloss.pdf}
    \caption{}
\end{subfigure}
\caption{Effects of applying MAGIC to a simulated dataset containing four well-separated subpopulations of cells.
(a) Plot of the first two PCs, created using log-normalized expression values.
Each point represents a cell and is coloured green (subpopulations $A$ and $C$) or orange ($B$ or $D$).
$A$ and $B$ are represented by open circles while $C$ and $D$ are represented by crosses.
The percentage of variance explained by each PC is shown in parentheses.
(b) Same as (a) but using the denoised expresssion values from MAGIC.
(c) Differences in log-expression with respect to the primary or secondary separations.
Each point represents a gene that is differentially expressed (blue) or not (grey).
(d) Same as (c) but for denoised expression values from MAGIC.}
\label{fig:fourclusters}
\end{figure}

Finally, we examined the consequences of using the denoised expression values in a more quantitative analysis.
We simulated a data set with two well-separated subpopulations and computed $p$-values for the genes that were not differentially expressed (DE) between those subpopulations.
As the null hypothesis is true, we expect that the distribution of $p$-values is uniform.
This is indeed the case when the log-transformed normalized expression values are used (Figure~\ref{fig:twoclusters}a).
However, the $p$-values computed from the denoised expression values are strongly skewed towards zero (Figure~\ref{fig:twoclusters}b).
This is a consequence of sharing information between neighbouring cells, which squeezes the expression values for all cells towards their subpopulation means.
As a result, the testing framework cannot properly account for the estimation uncertainty of the subpopulation-specific sample means.
This leads to loss of type I error control and false positives, i.e., detection of all non-DE genes as being significantly DE.

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=2]{../scripts/pics/fakede.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{../scripts/pics/fakede.pdf}
    \caption{}
\end{subfigure}
\caption{Distribution of $p$-values for non-DE genes, 
in a comparison between two simulated groups of cells on (a) log-normalized expression values or (b) denoised expression values from MAGIC.
All $p$-values were computed using the Wilcoxon rank-sum test.}
\label{fig:twoclusters}
\end{figure}

To demonstrate that our concerns are not purely academic, we applied MAGIC to two publicly available data sets from 10X Genomics \cite{zheng2017massively}.
The first data set involves only ERCC spike-in transcripts in gel beads or empty droplets.
When a $t$-stochastic neighbour embedding ($t$-SNE) \cite{van2008visualizing} plot was constructed on the log-normalized expression values, the gel beads and empty droplets formed distinct clusters distinguished by library size (Figure~\ref{fig:realdata}a), as expected.
However, the application of MAGIC produced a trajectory between the beads and the empty droplets (Figure~\ref{fig:realdata}b).
This is attributable to the data diffusion procedure where information is shared between otherwise distinct clusters.
The second data set contained a homogenous population of cells from the 293T cell line (Figure~\ref{fig:realdata}c).
Here, MAGIC amplified the separation between two subpopulations of cells that were otherwise weakly separated,
and created clear linear trajectories within each of those subpopulations (Figure~\ref{fig:realdata}d).
The most generous interpretation of these results would be that there was some underlying biology in each population that was not recovered with other analysis methods;
the most critical interpretation would be that MAGIC is creating artificial structure in a homogeneous population.

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=2]{../scripts/pics/ercc_results.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{../scripts/pics/ercc_results.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=2]{../scripts/pics/293t_results.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{../scripts/pics/293t_results.pdf}
    \caption{}
\end{subfigure}
\caption{Dimensionality reduction results after applying MAGIC to real scRNA-seq data.
(a, b) $t$-SNE plots of the ERCC data set (a) before and (b) after applying MAGIC.
Each point represents a cell or a cell-equivalent library, coloured according to the library size (purple for lowest, yellow for highest).
(c, d) PCA plots of the 293T data set (c) before and (d) after MAGIC.
Each point represents a cell, coloured by separation along PC1 in (d).
The percentage of variance explained by each PC is shown in parentheses.}
\label{fig:realdata}
\end{figure}

The same criticisms apply to real data analyses at the gene level.
We combined two separate scRNA-seq data sets involving B cells and CD4\textsuperscript{+} T helper cells \cite{zheng2017massively}.
After applying MAGIC, we examined the correlation between the denoised expression values for CD19 (a B cell marker) and CD4. 
We observed a strong negative correlation across all cells (Figure~\ref{fig:tb}), as expected.
However, we also observed strong negative correlations between these two markers within each cell type (Figure~\ref{fig:tb}a).
This is misleading as it implies co-regulation between CD19, which is not expressed in T cells; and CD4, which is not expressed in B cells.
Moreover, increasing the number of cells and re-running MAGIC alters the sign of these within-cell type correlations (Figure~\ref{fig:tb}b).
We speculate that this instability arises from different modes of data diffusion, where intra-population factors such as library size variation become more prominent with a greater number of cells.
Thus, not only does MAGIC yield results that are highly unlikely from a biological perspective, these results are not consistent with respect to the size of the data set.

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{../scripts/pics/tb_results.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=2]{../scripts/pics/tb_results.pdf}
    \caption{}
\end{subfigure}
\caption{Denoised expression values for CD19 and CD4 for a combined data set of B and T helper cells, after subsampling to (a) 200 or (b) 5000 cells of each type.}
\label{fig:tb}
\end{figure}

Here, we have described several scenarios in which the application of MAGIC yields a distorted representation of the data.
This is primarily driven by the use of data diffusion, though some of these issues may generalize to other denoising or imputation approaches.
We recommend treating results generated by MAGIC with caution, especially if they are not reproducible with conventional analysis strategies.

\section*{Methods}

\subsubsection*{Running MAGIC}
Given a count matrix, MAGIC was run on the square root of the library size-normalized expression values, as recommended at \url{https://github.com/KrishnaswamyLab/MAGIC}.
We set the diffusion time $t=10$ for consistency across all usage scenarios, though similar results were usually obtained with the automatically determined $t$.
Rmagic v1.0.0 with magic v1.2.1 was used for all analyses, running on R v3.5.0 and Python v3.5.2.

\subsubsection*{Simulation with no structure}
For the simulation used in Figure~\ref{fig:nostructure}, we considered an toy experiment containing 1000 genes and 500 cells.
The count for gene $g$ in cell $i$ was independently sampled from a Poisson distribution with mean $\lambda_{gi}$.
We defined $\lambda_{gi}= \lambda_g s_i$ for some size factor $s_i$.
Values of $\lambda_g$ were sampled from a $\mbox{Uniform}(0, 10)$ distribution, while $s_i$ was sampled from a $\mbox{Uniform}(1, 2)$ distribution.       
MAGIC was run on the count matrix as described above, and the denoised expression values were used directly for PCA.
We also computed Pearson's correlation between the denoised expression profiles for each pair of genes, and visualized the pair with the largest absolute correlation.
For comparison, these analyses were repeated using log-transformed normalized expression values.

\subsubsection*{Simulation with primary and secondary separation}
For the simulation used in Figure~\ref{fig:fourclusters}, we again considered an experiment containing 1000 genes and 500 cells.
Genes were split into a non-DE set (800 genes), a primary DE set (100) and a secondary DE set (100).
Cells were split into four equally sized subpopulations $A$, $B$, $C$ and $D$.
In the non-DE set, counts for gene $g$ were independently sampled from a Poisson distribution with mean $\lambda_{g}$, itself sampled as described above.
The primary DE set was further halved into upregulated and downregulated subsets.
In the upregulated primary DE set, counts were sampled from a $\mbox{Poisson}(5)$ distribution for subpopulations $A$ and $B$, 
and from a $\mbox{Poisson}(1)$ distribution for $C$ and $D$; and vice versa for the downregulated primary set.
The secondary DE set was similarly halved into upregulated and downregulated subsets.
In the upregulated secondary set, counts were sampled from a $\mbox{Poisson}(2)$ distribution for subpopulations $A$ and $C$, 
and from a $\mbox{Poisson}(1)$ distribution for $B$ and $D$; and vice versa for the downregulated primary set.
This scheme ensures that the DE genes are balanced and avoids introducing composition biases \cite{robinson2010scaling} that would complicate normalization.

MAGIC was run on the simulated count matrix as previously described.
PCA was performed on the denoised expression values to examine the separation between the known subpopulations.
We also examined the differences in denoised expression between subpopulations for the non-DE, primary and secondary DE gene sets.
For comparison, all analyses were repeated using log$_2$-transformed normalized expression values.

\subsubsection*{Simulation with two well-separated subpopulations}
For the simulation used in Figure~\ref{fig:twoclusters}, we considered an experiment containing 10000 genes and 500 cells.
Genes were split into a non-DE set (9900 genes) and a DE set (100).
Cells were split into two equally sized subpopulations $A$ and $B$.
In the non-DE set, counts for gene $g$ were independently sampled from a Poisson distribution with mean $\lambda_{g}$ as described above.
The DE set was further halved into upregulated and downregulated subsets.
In the upregulated DE set, counts were sampled from a $\mbox{Poisson}(5)$ distribution for subpopulations $A$ and $B$, and set to zero for $C$ and $D$;
and vice versa for the downregulated set.
MAGIC was run on the simulated count matrix and we tested for differences between the denoised expression values of each group using the Wilcoxon rank-sum test.
The $p$-values for all non-DE genes were used to construct a histogram.
For comparison, we also computed $p$-values from the log$_2$-normalized expression.

\subsubsection*{Real data analyses}
The ERCC spike-in data set was obtained from the 10X Genomics website (\url{https://support.10xgenomics.com/single-cell-gene-expression/datasets/}) as a raw count matrix.
Libraries with total counts less than or equal to 100 were removed prior to further analysis.
Spike-in transcripts with average counts below 0.01 were similarly removed.
MAGIC was run on the (root-normalized) count matrices as described previously.
A $t$-SNE plot was generated using the Rtsne package v0.13 (\url{https://cran.r-project.org/package=Rtsne}) with a perplexity of 40 and no initial PCA
(due to the relatively small number of features in the filtered spike-in set).
For comparison, we also generated a $t$-SNE plot from the log-normalized expression.

The 293T data set was obtained from the 10X Genomics website as a filtered count matrix.
Genes with average counts below 0.01 were removed prior to further analyses.
MAGIC was run on the (root-normalized) count matrices as described previously.
An approximate PCA was performed on the denoised matrix using the irlba package v2.3.2 (\url{https://cran.r-project.org/package=irlba}) to examine the first two PCs.
For comparison, this was repeated using the log-normalized expression values.

The B and CD4\textsuperscript{+} T helper cell data sets were obtained from the 10X Genomics website as filtered count matrices.
We randomly sampled 200 or 5000 cells from each data set, combined the count matrices and filtered out genes with average counts below 0.01.
We ran MAGIC on the (root-normalized) count matrices as described previously, to obtain denoised expression values for further inspection.
We repeated this 10 times with different samplings and used a representative result in Figure~\ref{fig:tb}.

\bibliography{ref}
\bibliographystyle{unsrt}


\end{document}
