\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in,marginparwidth=2in]{geometry}

% use Unicode characters - try changing the option if you run into troubles with special characters (e.g. umlauts)
\usepackage[utf8]{inputenc}

% clean citations
\usepackage{cite}

% hyperref makes references clicky. use \url{www.example.com} or \href{www.example.com}{description} to add a clicky url
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% improves typesetting in LaTeX
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% text layout - change as needed
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in 
\textheight 8.75in

% Remove % for double line spacing
%\usepackage{setspace} 
%\doublespacing

% use adjustwidth environment to exceed text width (see examples in text)
\usepackage{changepage}

% adjust caption style
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,singlelinecheck=off]{caption}

% remove brackets from references
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% headrule, footrule and page numbers
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}

% use \textcolor{color}{text} for colored text (e.g. highlight to-do areas)
\usepackage{color}

% define custom colors (this one is for figure captions)
\definecolor{Gray}{gray}{.25}

% this is required to include graphics
\usepackage{graphicx}

% use if you want to put caption to the side of the figure - see example in text
\usepackage{sidecap}

% use for have text wrap around figures
\usepackage{wrapfig}
\usepackage[pscoord]{eso-pic}
\usepackage[fulladjust]{marginnote}
\reversemarginpar

% Adding multirow.
\usepackage{multirow}

% Other required things:
\usepackage{color}
\usepackage{subcaption}
\captionsetup[subfigure]{justification=centering}

% document begins here
\begin{document}
\vspace*{0.35in}

% title goes here:
\begin{flushleft}
{\Large
    \textbf\newline{Too much magic: avoiding illusions in single cell RNA sequencing data analysis}
}
\newline

% authors go here:
%\\
Timothy Curtis\textsuperscript{1,*}
\\
\bigskip
\bf{1} Diagon Institute of Computational Biology, London, United Kingdom \\
\bigskip

\end{flushleft}

\noindent
A new imputation method ``MAGIC" \cite{vandijk2018recovering} has recently been proposed for denoising single-cell RNA sequencing (scRNA-seq) data.
This method involves constructing an affinity matrix based on the distances between cells in high-dimensional space;
applying a diffusion process to strengthen similarities between cells due to existing trends in the data;
and multiplying the exponentiated affinity matrix with the original count matrix to obtain denoised expression values for downstream analysis.
The authors claim that the use of MAGIC removes technical noise and improves the biological signal in scRNA-seq data.
Indeed, the noisy nature of scRNA-seq data has been documented \cite{grun2015design} and many computational methods have been developed to address this problem \cite{bacher2016design}.
Unfortunately - as we will argue below - using MAGIC is not the solution.

The raison d'\^etre of the MAGIC procedure is the use of the diffusion process to reinforce existing trends.
The affinity matrix is exponentiated to represent the probability of one cell reaching another via a random walk of length $t$, possibly involving a number of intermediate cells.
This procedure systematically increases the affinities between pairs of cells that are connected by many close neighbours.
Conversely, affinities are downweighted between pairs of cells in sparse regions of the high dimensional space, 
where sporadic similarities are more likely to be caused by technical noise or outliers.
By performing diffusion, MAGIC can achieve aggressive denoising compared to other methods like $k$-NN imputation.
However, this also runs the risk of distorting the biological structure in the denoised expression matrix.

This effect is best demonstrated with a simple simulation.
We applied MAGIC to a simulated matrix of Poisson-distributed counts that contained no structure other than modest differences in library size across cells.
Upon performing a principal components analysis (PCA) on the denoised expression matrix, we observed a ``trajectory'' of cells along the first principal component (PC) (Figure~\ref{fig:nostructure}a) that was not present with the original expression matrix (Figure~\ref{fig:nostructure}b).
This trajectory is a manifestation of spurious structure, likely arising from stochastic differences in the affinity matrix that are inflated during diffusion.
These initial differences in affinity are probably driven by differences in variance when library sizes are variable across cells, assuming that mean differences are eliminated by library size normalization.
While the creation of artificial structure is obviously undesirable, the interpretation of denoised expression values at the gene level is even more compromised.
Independent genes now appear to be strongly co-regulated (Figure~\ref{fig:nostructure}c) whereas no such effect is observed in the original data (Figure~\ref{fig:nostructure}d).

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 15mm,clip,page=1]{scripts/pics/nostructure_with_magic.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 15mm,clip,page=1]{scripts/pics/nostructure_without_magic.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 15mm,clip,page=2]{scripts/pics/nostructure_with_magic.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 15mm,clip,page=2]{scripts/pics/nostructure_without_magic.pdf}
    \caption{}
\end{subfigure}
\caption{Effects of applying MAGIC to a simulated dataset containing independently sampled Poisson-distributed counts with no structure.
(a) Plot of the first two PCs after performing PCA on the denoised expression matrix.
Each point represents a cell, coloured according to its library size (purple-green-yellow gradient, from lowest to highest).
The percentage of variance explained by each PC is shown in parentheses.
(b) Same as (a) but on the log-transformed normalized counts without running MAGIC.
(c) Denoised expression values of the two genes with the highest pairwise correlation after running MAGIC.
Each point represents a cell.
The line of best fit and R$^2$ are also shown in red.
(d) Same as (d) but on log-expression values without running MAGIC.}
\label{fig:nostructure}
\end{figure}

One might dismiss the previous simulation as being somewhat pathological.
After all, there are no biological systems that exhibit no structure whatsoever.
Perhaps a more pertinent question would be: how does MAGIC perform in the presence of genuine structure? 
To explore this, we performed another simulation involving four equally sized subpopulations of cells, denoted here as $A$, $B$, $C$ and $D$.
The primary separation occurs between $\{A, B\}$ and $\{C, D\}$ and the secondary separation occurs between $A/B$ and $C/D$.
Application of MAGIC eliminated the secondary separation (Figure~\ref{fig:fourclusters}a), merging $A$ with $B$ and $C$ with $D$.
This is again a consequence of the diffusion process, where weak similarities between cells in $A$ and $B$ (or $C$ and $D$) are increasingly strengthened over diffusion time.
More information is subsequently shared between those pairs of subpopulations during the denoising step.
This squeezes the subpopulations together and eliminates geniune structure that is otherwise clearly visible (Figure~\ref{fig:fourclusters}b).
Gene-level interpretation is also compromised as differences in expression are shrunk towards zero for the subset of genes responsible for the secondary separation (Figure~\ref{fig:fourclusters}c, d).

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{scripts/pics/minor_with_magic.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{scripts/pics/minor_without_magic.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=2]{scripts/pics/minor_with_magic.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=2]{scripts/pics/minor_without_magic.pdf}
    \caption{}
\end{subfigure}
\caption{Effects of applying MAGIC to a simulated dataset containing four well-separated subpopulations of cells.
(a) Plot of the first two PCs, created using the denoised expression values from MAGIC.
Each point represents a cell and is coloured green (subpopulations $A$ and $C$) or orange ($B$ or $D$).
$A$ and $B$ are represented by open circles while $C$ and $D$ are represented by crosses.
The percentage of variance explained by each PC is shown in parentheses.
(b) Same as (a) but using the original counts.
(c) Differences in denoised expression with respect to the primary or secondary separations.
Each point represents a gene and is coloured according to whether it is DE (blue) or not (grey).
(d) Same as (c) but for log-fold changes computed from the original counts.}
\label{fig:fourclusters}
\end{figure}

Finally, we inspected the gene-wise denoised expression profiles in another simulation involving two well-separated subpopulations.
As expected, we observed strong separation between the two subpopulations regardless of whether MAGIC was used (Figure~\ref{fig:twoclusters}).
However, MAGIC also introduced spurious differences in expression between the two subpopulations among the non-DE genes (Figure~\ref{fig:twoclusters}a).
This is a consequence of sharing information between neighbouring cells, which squeezes the expression values for all cells towards their subpopulation means.
Estimation uncertainty will yield a different sample mean for each subpopulation, even for non-DE genes.
These small differences are amplified by aggressive denoising, manifesting as in strong differences in the distribution of expression values betweeen subpopulations.
One could argue that this spurious DE pattern is not of concern as the log-fold changes for the genes involved are relatively small (less than 0.2 in most cases).
However, this attitude is not consistent with the intention behind the use of denoising in the first place, 
i.e., to recover subtle biological changes that would otherwise be masked by technical noise.

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{scripts/pics/clusters_with_magic.png}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{scripts/pics/clusters_without_magic.png}
    \caption{}
\end{subfigure}
\caption{Heatmaps of (a) denoised expression values after running MAGIC and (b) log$_2$-expression values without MAGIC, 
in a simulation containing two well-separated subpopulations of cells.
Each column represents a cell and is labelled according to the subpopulation (i.e., cluster) to which it belongs.
Each row represents a gene, sorted by the difference between subpopulations and labelled according to whether it is truly DE.
Expression values are mean-centered along each row and capped at 0.2 for visibility.}
\label{fig:twoclusters}
\end{figure}

To demonstrate that our concerns are not purely academic, we applied MAGIC to two publicly available data sets from 10X Genomics \cite{zheng2017massively}.
The first data set contains only ERCC spike-in transcripts without any cells.
After using MAGIC, the libraries separated into two groups on a $t$-stochastic neighbour embedding \cite{van2008visualizing} plot (Figure~\ref{fig:realdata}a).
No such separation was observed in the original data (Figure~\ref{fig:realdata}b), which is more expected given the fixed composition of the ERCC spike-in mixture.
The second data set contained a homogenous population of cells from the 293T cell line.
Here, MAGIC amplified the separation between two subpopulations of cells that were otherwise weakly separated (Figure~\ref{fig:realdata}c),
and created clear linear trajectories within each of those subpopulations.
The generation of such strong signal is not consistent with a homogeneous population and is not observed in the original data (Figure~\ref{fig:realdata}d).
The most generous interpretation of these results would be that there was some underlying biology in each population that was not recovered with other analysis methods;
the most critical interpretation would be that MAGIC is creating artificial structure.

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{scripts/pics/ercc_results.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=2]{scripts/pics/ercc_results.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=1]{scripts/pics/293t_results.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,page=2]{scripts/pics/293t_results.pdf}
    \caption{}
\end{subfigure}
\caption{Plots of the first two PCs for the ERCC (a, b) and 293T (c, d) data sets from 10X Genomics, with (a, c) and without using MAGIC (b, d).
Each point represents a cell and the percentage of variance explained by each PC is shown in parentheses.
For each data set, cells in the two putative subpopulations are distinguished by colour.}
\label{fig:realdata}
\end{figure}

Here, we have described several scenarios in which the application of MAGIC yields a distorted representation of the data.
This is primarily driven by the use of data diffusion, though some of these issues may generalize to other denoising or imputation approaches.
We recommend treating results generated by MAGIC with caution, especially if they are not reproducible with conventional analysis strategies.

\section*{Methods}

\subsubsection*{Running MAGIC}
Given a count matrix, MAGIC was run on the square root of the library size-normalized counts, as recommended at \url{https://github.com/KrishnaswamyLab/MAGIC}.
We set the diffusion time $t=10$ for consistency across all usage scenarios, though similar results were usually obtained with the automatically determined $t$.
Rmagic v1.0.0 with magic v1.1.0 was used for all analyses, running on R v3.5.0 and Python v3.5.2.

\subsubsection*{Simulation with no structure}
For the simulation used in Figure~\ref{fig:nostructure}, we considered an toy experiment containing 1000 genes and 500 cells.
The count for gene $g$ in cell $i$ was independently sampled from a Poisson distribution with mean $\lambda_{gi}$.
We defined $\lambda_{gi}= \lambda_g s_i$ for some size factor $s_i$.
Values of $\lambda_g$ were sampled from a $\mbox{Uniform}(0, 10)$ distribution, while $s_i$ was sampled from a $\mbox{Uniform}(1, 2)$ distribution.       
MAGIC was run on the count matrix as described above, and the denoised expression values were used directly for PCA.
We also computed Pearson's correlation between the denoised expression profiles for each pair of genes, and visualized the pair with the largest absolute correlation.
For comparison, these analyses were repeated using log$_2$-transformed library size-normalized counts.

\subsubsection*{Simulation with primary and secondary separation}
For the simulation used in Figure~\ref{fig:fourclusters}, we again considered an experiment containing 1000 genes and 500 cells.
Genes were split into a non-DE set (800 genes), a primary DE set (100) and a secondary DE set (100).
Cells were split into four equally sized subpopulations $A$, $B$, $C$ and $D$.
In the non-DE set, counts for gene $g$ were independently sampled from a Poisson distribution with mean $\lambda_{g}$ which was sampled as described above.
The primary DE set was further halved into upregulated and downregulated subsets.
In the upregulated primary DE set, counts were sampled from a $\mbox{Poisson}(5)$ distribution for subpopulations $A$ and $B$, 
and from a $\mbox{Poisson}(1)$ distribution for $C$ and $D$; and vice versa for the downregulated primary set.
The secondary DE set was similarly halved into upregulated and downregulated subsets.
In the upregulated secondary set, counts were sampled from a $\mbox{Poisson}(2)$ distribution for subpopulations $A$ and $C$, 
and from a $\mbox{Poisson}(1)$ distribution for $B$ and $D$; and vice versa for the downregulated primary set.
This scheme ensures that the DE genes are balanced and avoids introducing composition biases \cite{robinson2010scaling} that would complicate normalization.

MAGIC was run on the simulated count matrix as previously described.
PCA was performed on the denoised expression values to examine the separation between the known subpopulations.
We also examined the differences in denoised expression between subpopulations for the non-DE, primary and secondary DE gene sets.
For comparison, all analyses were repeated using log$_2$-transformed normalized counts.

\subsubsection*{Simulation with two well-separated subpopulations}
For the simulation used in Figure~\ref{fig:twoclusters}, we again considered an experiment containing 1000 genes and 500 cells.
Genes were split into a non-DE set (900 genes) and a DE set (100).
Cells were split into two equally sized subpopulations $A$ and $B$.
In the non-DE set, counts for gene $g$ were independently sampled from a Poisson distribution with mean $\lambda_{g}$ which was sampled as described above.
The DE set was further halved into upregulated and downregulated subsets.
In the upregulated DE set, counts were sampled from a $\mbox{Poisson}(5)$ distribution for subpopulations $A$ and $B$, and set to zero for $C$ and $D$;
and vice versa for the downregulated set.
MAGIC was run on the simulated count matrix and the denoised expression values were used to create a heatmap using the pheatmap package v1.0.10 (\url{https://cran.r-project.org/package=pheatmap}).
For comparison, an equivalent heatmap was also constructed using the log$_2$-normalized counts.

\subsubsection*{Real data analyses}
The 293T and ERCC data sets were obtained from the 10X Genomics website (\url{https://support.10xgenomics.com/single-cell-gene-expression/datasets/}) as filtered count matrices.
Genes with average counts below 0.1 were removed prior to further analyses.
MAGIC was run on the (root-normalized) count matrices as described previously.
For the 293T data set, an approximate PCA was performed on the denoised matrix using the irlba package v2.3.2 (\url{https://cran.r-project.org/package=irlba}) to examine the first two PCs.
For the ERCC data set, a $t$-SNE plot was generated using the Rtsne package v0.13 (\url{https://cran.r-project.org/package=Rtsne}) with a perplexity of 90 and no initial PCA
(due to the relatively small number of spike-in features).
We used $t$-SNE here instead of PCA to highlight subtle artifacts introduced by MAGIC that are not apparent on a PCA plot.
In each data set, putative subpopulations were manually identified and highlighted for visualization purposes.

\bibliography{ref}
\bibliographystyle{unsrt}


\end{document}
